File: src/java/magma/api/collect/list/List.java
package magma.api.collect.list;

import magma.api.option.Option;

public interface List<T> extends Iterable<T> {
    List<T> addLast(T element);

    int size();

    Option<List<T>> subList(int startInclusive, int endExclusive);

    Option<T> findLast();

    Option<T> findFirst();

    Option<T> find(int index);

    List<T> addAll(Iterable<T> others);

    boolean contains(T element);

    List<T> addFirst(T element);

    boolean isEmpty();

    List<T> removeValue(T element);

    Option<List<T>> removeLast();
}

File: src/windows/magma/api/collect/list/List.c
#include "./List.h"
export interface List<T> {
	addLast(element: T): List<T>;
	query(): Query<T>;
	size(): number;
	subList(startInclusive: number, endExclusive: number): Option<List<T>>;
	findLast(): Option<T>;
	findFirst(): Option<T>;
	find(index: number): Option<T>;
	queryWithIndices(): Query<Tuple2<number, T>>;
	addAll(others: List<T>): List<T>;
	contains(element: T, equator: (arg0 : T, arg1 : T) => Bool): Bool;
	queryReversed(): Query<T>;
	addFirst(element: T): List<T>;
	isEmpty(): Bool;
	equalsTo(other: List<T>, equator: (arg0 : T, arg1 : T) => Bool): Bool;
	removeValue(element: T, equator: (arg0 : T, arg1 : T) => Bool): List<T>;
	removeLast(): Option<List<T>>;
}
